clustercontroller-reindexer/src/main/java/ai/vespa/reindexing/Reindexer.java

```diff
@@ -0,0 +1,216 @@
+import com.yahoo.document.Field;
+import com.yahoo.document.select.parser.ParseException;
+import com.yahoo.documentapi.DocumentAccess;
+import com.yahoo.documentapi.ProgressToken;
+import com.yahoo.documentapi.VisitorControlHandler;
+import com.yahoo.documentapi.VisitorControlHandler.CompletionCode;
+import com.yahoo.documentapi.VisitorParameters;
+import com.yahoo.documentapi.VisitorSession;
+import com.yahoo.documentapi.messagebus.protocol.DocumentProtocol;
+import com.yahoo.vespa.curator.Lock;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import static com.yahoo.documentapi.VisitorControlHandler.CompletionCode.ABORTED;
+import static java.util.Objects.requireNonNull;
+import static java.util.logging.Level.FINE;
+import static java.util.logging.Level.INFO;
+import static java.util.logging.Level.WARNING;
+import static java.util.stream.Collectors.joining;
+
+/**
+ * Progresses reindexing efforts by creating visitor sessions against its own content cluster,
+ * which send documents straight to storage — via indexing if the documenet type has "index" mode.
+ * The {@link #reindex} method blocks until unterrupted, or util no more reindexing is left to do.
+ *
+ * @author jonmv
+ */
+public class Reindexer {
+
+    private static final Logger log = Logger.getLogger(Reindexer.class.getName());
+
+    private final Cluster cluster;
+    private final Map<DocumentType, Instant> ready;
+    private final ReindexingCurator database;
+    private final DocumentAccess access;
+    private final Clock clock;
+
+    public Reindexer(Cluster cluster, Map<DocumentType, Instant> ready, ReindexingCurator database,
+                     DocumentAccess access, Clock clock) {
+        for (DocumentType type : ready.keySet())
+            cluster.bucketOf(type); // Verifies this is known.
+
+        this.cluster = cluster;
+        this.ready = new TreeMap<>(ready);
+        this.database = database;
+        this.access = access;
+        this.clock = clock;
+    }
+
+    /** Starts and tracks reprocessing of ready document types until done, or interrupted. */
+    public void reindex() throws ReindexingLockException {
+        try (Lock lock = database.lockReindexing()) {
+            Reindexing reindexing = database.readReindexing();
+            for (DocumentType type : ready.keySet()) { // We consider only document types for which we have config.
+                if (ready.get(type).isAfter(clock.instant())) {
+                    log.log(WARNING, "Received config for reindexing which is ready in the future — will process later " +
+                                     "(" + ready.get(type) + " is after " + clock.instant() + ")");
+                }
+                else {
+                    // If this is a new document type (or a new cluster), no reindexing is required.
```

Comment:
This will intentionally happen for _all the things_ the first time this all rolls out to existing deployments, so no reprocessing will start immediately.